pennylane/data/data_manager/foldermap.py

```diff
@@ -0,0 +1,177 @@
+        matches ``params``."""
+
+        try:
+            data_names_to_params = self.__curr_level["__params"]
+        except KeyError as exc:
+            raise RuntimeError("Can only call find() from top level of foldermap") from exc
+
+        try:
+            param_names: List[str] = data_names_to_params[data_name]
+        except KeyError as exc:
+            raise ValueError(f"No datasets with data name: '{data_name}'") from exc
+
+        curr: List[Tuple[Description, Union[FolderMapView, DataPath]]] = [
+            (Description(()), self[data_name])
+        ]
+        todo: List[Tuple[Description, Union[FolderMapView, DataPath]]] = []
+        done: List[Tuple[Description, DataPath]] = []
+
+        for param_name in param_names:
+            param_arg = params.get(param_name, missing_default)
+
+            while curr:
+                curr_description, curr_level = curr.pop()
+                if param_arg == ParamArg.FULL:
+                    next_params = curr_level
+                elif param_arg == ParamArg.DEFAULT:
+                    default = curr_level.get_default_key()
+                    if default is None:
+                        raise ValueError(f"No default available for parameter '{param_name}'")
+                    next_params = (default,)
+                elif isinstance(param_arg, str):
+                    next_params = (param_arg,)
+                else:
+                    next_params = param_arg
+
+                todo.extend(
+                    (
+                        Description((*curr_description.items(), (param_name, next_param))),
+                        curr_level[next_param],
+                    )
+                    for next_param in next_params
+                )
+
+            curr, todo = todo, curr
+
+        done.extend(curr)
+
+        return done
+
+    def __getitem__(
+        self, __key: Union[str, Literal[ParamArg.DEFAULT]]
+    ) -> Union["FolderMapView", DataPath]:
+        """Gets the item with key. If key is ``ParamArg.DEFAULT``, return the
+        item under the default parameter, or raise a ``ValueError`` if no
+        default exists."""
+        if __key in self.__PRIVATE_KEYS:
+            raise KeyError(__key)
+
+        if __key == ParamArg.DEFAULT:
+            default = self.get_default_key()
+            if default is None:
+                raise ValueError("No default available")
+            return self[default]
+
+        elem = self.__curr_level[__key]
+        if isinstance(elem, Mapping):
+            return FolderMapView(elem)
+
+        return DataPath(elem)
+
+    def __iter__(self) -> typing.Iterator[str]:
+        return (key for key in self.__curr_level if key not in self.__PRIVATE_KEYS)
+
+    def __len__(self) -> int:
+        return sum(1 for _ in self.__iter__())
+
+    def __repr__(self) -> str:
+        items_repr = ", ".join((f"{repr(k)}: {repr(v)}") for k, v in self.items())
```

Comment:
Again, why can't we do `repr(dict(self.items()))` here?

Reply:
good point